### PR TITLE
Fix to  map jumping when bounds cross longitude 180

### LINKED ISSUE #6985: How to constrain the map to a region spanning the 180/-180 longitude?
### Question

Mapbox-GL JS library allows using the `maxBounds` attribute or `setMaxBounds` function to constrain the displayed map to a specific geographic region. However, when the region spans the longitude of 180/-180 degrees, the result is the whole length of the “world strip” (between the two specified latitudes), instead of the desired box.

For example, the bounding box for Kiribati is {"northeast": [177.048, 4.900], "southwest": [-174.743, -11.646]} but the map cannot be constrained to this box correctly when using these coordinates directly.

(I realize this sounds like a specific use case, but I do think this warrants a clarification in the documentation.)

### Links to related documentation

https://www.mapbox.com/mapbox-gl-js/api/#map (parameter `options.maxBounds`)
https://www.mapbox.com/mapbox.js/api/v3.1.1/l-map-class/#map-maxbounds (map state `maxBounds` option)

### LINKED ISSUE #10447: Map incorrectly jumps if maxBounds crosses the longitude 180
**mapbox-gl-js version**: Latest (v2.1.1)

**browser**: Safari v14.0.3, also reproducible in other browsers

### Steps to Trigger Behavior

 1. Set maxBounds to [[158, -55], [190, -23]], this covers New Zealand and the Chatham Islands
 2. Zoom in to at least zoom level 5
 3. Try to pan right (to the east)

### Link to Demonstration

https://jsbin.com/lakamahoni/1/edit?html,output

### Expected Behavior

Map should let me pan to the east

### Actual Behavior

Once going east a certain point (presumably longitude 180), the map suddenly jumps to the (opposite) western most side of the maxBounds.

### PR DESCRIPTION
**Pull request by SnailBones**
## Launch Checklist
Closes #10447.
Closes #6985.

These two issues arise from the same cause: when maxBounds crossed the 180th Meridian, the view would be adjusted incorrectly, snapping to the other edge of the map and preventing the user from crossing the Meridian in either direction.

This PR provides a fix by wrapping and normalizing the bounds and position before applying the update.

To allow maps across the 180th meridian to work with correct (unwrapped) longitude, I added a check in setMaxBounds to unwrap them.

Two minor issues unsolved by this fix:

- The bounds span 360 degrees AND The map is moved quickly against the edge so that the center passes beyond the edge. The user can effectively "force" the map to the other side. (Previously this wouldn't have happened, but the only working case was bounds from -180 to 180.)
- Bounds spanning over 360 degrees are treated at higher zoom levels as effectively no bounds. (I think this isn't really a valid use case, and I'm not seeing any Github issues about it.  If I'm wrong, I think it's still an improvement over previous/current behavior which is sudden snapping to an unrelated point on the map.)

These both happen because the map center is now "wrapped" before checking maxBounds. This prevents issues with the 180th meridian, but it also means that the map position has no reference for where it is on the map separate from its world position.

 - [x] Fix crossing the antimeridian from west to east
 - [x] Fix crossing the antimeridian from east to west
 - [x] write tests for all new functionality
 - [x] Check edge cases
 - [x] Add support for entering bounds without unwrapping longitudes (As described in #6985)
 - [ ] post benchmark scores
 - [x] manually test the debug page
 - [ ] tagged `@mapbox/gl-native` if this PR includes shader changes or needs a native port
 - [x] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog>'maxBounds' property across the 180th meridian work correctly.</changelog>`

### CODE REVIEW

File: src/geo/transform.js [hunk 1]
```diff
@@ -1355,11 +1361,9 @@ class Transform {
 
         this._constraining = true;
 
-        let minY = -90;
-        let maxY = 90;
-        let minX = -180;
-        let maxX = 180;
-        let sy, sx, x2, y2;
+        let minY = 0;
```
Comment (SnailBones) [inline src/geo/transform.js:1364]: Removing these initial values since they are never used (and suggest incorrect lat/long units).

**Review by rreusser — COMMENTED** [review]
Still need to work through the behavior more carefully. Submitting partial comments from a not-yet-finished review.

File: src/geo/transform.js [hunk 2]
```diff
@@ -887,17 +887,20 @@ class Transform {
     zoomScale(zoom: number) { return Math.pow(2, zoom); }
     scaleZoom(scale: number) { return Math.log(scale) / Math.LN2; }
 
+    // World coordinates from LngLat
```
[the remark this answers is not in the document]
Comment (SnailBones) [reply src/geo/transform.js:890]: Great idea!

File: test/unit/geo/transform.test.js [hunk 3]
```diff
@@ -117,6 +117,112 @@ test('transform', (t) => {
...
 
+    t.test('maxBounds should not jump to the wrong side when crossing 180th meridian (#10447)', (t) => {
+        t.test(' to the East', (t) => {
+            const transform = new Transform();
+            transform.zoom = 6;
+            transform.resize(500, 500);
+            transform.lngRange = [160, 190];
+            transform.latRange = [-55, -23];
+
+            transform.center = new LngLat(-170, -40);
```
[the remark this answers is not in the document]
Comment (SnailBones) [reply test/unit/geo/transform.test.js:128]: Yes, this test sets the center to the east bound in order to test that the map position snaps to the correct side.  The exact position is arbitrary, it could be any position out of bounds on the east side of the map.

File: src/geo/transform.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (SnailBones) [reply src/geo/transform.js:890]: How's this?
```suggestion
    // Transform from LngLat to world coordinates [-180, 180] x [-90, 90] --> [0, this.worldSize] x [0, this.worldSize]
```

File: src/geo/transform.js (hunk 1 above)
Comment (SnailBones) [reply src/geo/transform.js:1364]: Should `maxY = -Infinity`?

File: debug/antimeridian.html [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+<body>
+<div id='map'></div>
+
+<script src='../dist/mapbox-gl-dev.js'></script>
+<script src='../debug/access_token_generated.js'></script>
+
+<script>
+
+var map = new mapboxgl.Map({
```
Comment (rreusser) [inline debug/antimeridian.html:22]: For convenience (related to `map.transform.center` debug logs below, so that you can check this yourself if you need to fire up this page).

```suggestion
var map = window.map = new mapboxgl.Map({
```

File: debug/antimeridian.html [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    zoom: 6,
+    center: [179.012, -18.124],
+    maxBounds: [[175, -25], [-175, -10]]
+});
+map.transform.center = new mapboxgl.LngLat(190, -40);
+
+document.addEventListener('keypress', (e) => {
+    console.log(e.key);
+    if (e.key === " ") {
+        console.log(map.transform.center);
```
Comment (rreusser) [inline debug/antimeridian.html:34]:
```suggestion
        console.log('map.transform.center:', map.transform.center);
```

File: debug/antimeridian.html [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        console.log(map.transform.point);
```
Comment (rreusser) [inline debug/antimeridian.html:35]: For clarity:

```suggestion
        console.log('map.transform.point:', map.transform.point);
```

File: src/geo/transform.js [hunk 7]
```diff
@@ -1355,11 +1361,9 @@ class Transform {
 
         this._constraining = true;
 
-        let minY = -90;
-        let maxY = 90;
-        let minX = -180;
-        let maxX = 180;
-        let sy, sx, x2, y2;
+        let minY = Infinity;
```
Comment (rreusser) [inline src/geo/transform.js:1364]:
```suggestion
        let minY = 90;
```

File: src/geo/transform.js [hunk 8]
```diff
@@ -1355,11 +1361,9 @@ class Transform {
...
+        let maxY = -Infinity;
```
Comment (rreusser) [inline src/geo/transform.js:1365]: As opposed to longitude, which wraps, these could never even possibly be outside the range [-90, 90], right, so that they can be initialized with those bounds?

```suggestion
        let maxY = -90;
```

File: src/geo/transform.js [hunk 9]
```diff
@@ -1355,11 +1361,9 @@ class Transform {
...
+        let minX, maxX, sy, sx, y2;
```
Comment (rreusser) [inline src/geo/transform.js:1366]: It looks to me like minX and maxX could be accessed uninitialized. Do they need initialization?

```suggestion
        let minX = Infinity;
        let maxX = -Infinity;
        let sy, sx, y2;
```

**Review by rreusser — CHANGES_REQUESTED** [review]
It looks really good so far and seems to work well when I test it out! Nice work! 👏 👏  I had just a couple nitpicks as well as a question about what looks to me like possible use of uninitialized variables.

File: src/geo/transform.js (hunk 8 above)
Comment (SnailBones) [reply src/geo/transform.js:1365]: Not quite, these values are in pixel coordinates. (Those 90 and 180 values are a misleading!) I believe the correct range shouold be: `[0, this.worldSize]`. 

It makes no difference to functionality since the values are always modified before they are used, the only reason to initialize them is to prevent a Flow error. My first thought was that `[0, 0]` was the clearest way to show an unused value. But happy to change it to`[0, this.worldSize]` if you think that makes more sense!

Comment (rreusser) [reply src/geo/transform.js:1365]: Ah, got it! Nah, infinities are great then. Yes, I assumed it was straight latitude/longitude, of course. LGTM!

**Review by rreusser — APPROVED** [review]
Thanks for the clarification. I'm content here! I wonder if it really needs the long-term debug page, but no particular issue with it either.
